File: NeetCode/2-D_Dynamic_Programming/Unique_Paths_II/unique__paths__i_i.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        height = len(obstacleGrid)
        width = len(obstacleGrid[0])
        dp = [[0 for _ in range(width + 1)] for _ in range(height + 1)]

        dp[height - 1][width - 1] = 1
        for y in range(height - 1, -1, -1):
            for x in range(width - 1, -1, -1):
                if obstacleGrid[y][x] == 1:
                    continue
                dp[y][x] += dp[y + 1][x] + dp[y][x + 1]

        return dp[0][0]


if __name__ == "__main__":
    sol = Solution()
    obstacleGrid = [
        [0, 0, 0],
        [0, 0, 0],
        [0, 1, 0],
    ]
    print(sol.uniquePathsWithObstacles(obstacleGrid))
    print("Running Solution...")

File: NeetCode/2-D_Dynamic_Programming/Unique_Paths_II/Unique_Paths_II.cpp
// Unique_Paths_II.cpp
//
// Compile locally with:
// clang++ Unique_Paths_II.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int uniquePathsWithObstacles(vector<vector<int>> &obstacleGrid)
    {
        int height = obstacleGrid.size();
        int width = obstacleGrid[0].size();
        vector<vector<long>> dp(height + 1, vector<long>(width + 1, 0));
        if (obstacleGrid[height - 1][width - 1] == 0)
        {
            dp[height - 1][width - 1] = 1;
        }

        for (int y = height - 1; y >= 0; y--)
        {
            for (int x = width - 1; x >= 0; x--)
            {
                if (obstacleGrid[y][x] == 1)
                {
                    continue;
                }
                dp[y][x] += dp[y + 1][x] + dp[y][x + 1];
            }
        }
        return dp[0][0];
    }
};

int main()
{
    cout << "Running Unique_Paths_II.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
